src/imitation/data/types.py

```diff
@@ -109,15 +119,39 @@ def __post_init__(self):
                 "obs and acts must have same number of timesteps: "
                 f"{len(self.obs)} != {len(self.acts)}"
             )
-        if len(self.obs) == 0:
-            raise ValueError("Must have non-zero number of observations.")
 
         if self.infos is not None and len(self.infos) != len(self.obs):
             raise ValueError(
                 "obs and infos must have same number of timesteps: "
                 f"{len(self.obs)} != {len(self.infos)}"
             )
 
+    @overload
+    def __getitem__(self: T, key: slice) -> T:
+        pass
+
+    @overload
+    def __getitem__(self, key: int) -> Dict[str, np.ndarray]:
+        pass
+
+    def __getitem__(self, key):
+        """See TransitionsMinimal docstring for indexing and slicing semantics."""
+        # Extract items using `dataclasses.fields` + dict comprehension instead of using
+        # `dataclasses.asdict` because `asdict` expensively and undocumentedly
+        # deep-copies every numpy array value.
+        # See https://stackoverflow.com/a/52229565/1091722.
+        d = {f.name: getattr(self, f.name) for f in dataclasses.fields(self)}
+        d_item = {k: v[key] for k, v in d.items()}
+
+        if isinstance(key, slice):
+            # Return type is the same as this dataclass. Replace field value with
+            # slices.
+            return dataclasses.replace(self, **d_item)
+        else:
+            assert isinstance(key, int)
+            # Return type is a dictionary. Array values have no batch dimension.
+            return d_item
```

Comment:
It's not immediately clear here from the implementation, but the reason this returns dictionary is because `Dict[str, Union[np.ndarray, Tensor]]` is a good `torch.util.data.Dataset` return type.

Pytorch natively knows how to join together lots of dictionaries to build dictionary batches, which is what we agreed would be good for `BC` training to use. `torch.util.data.DataLoader` samples shuffled batches from a `Transitions` instance (which recall, now implements `torch...Dataset`) by picking random indices, and then joining together the elements associated with each index.

I'll expand the comment here to explain that dictionary is easy for `DataLoader` to work with.

Reply:
I looked more closely into Torch's `default_collate` and it seems like they know how to join individual NamedTuple as well into batches as well.

However for the sake of allowing folks to easily use any ducktyped data source (without manufacturing a named tuple) it seems like a good idea to use dict with keys "obs" and "acts" in `BC`.

Reply:
OK, this seems reasonable, although I agree it's worth documenting the design rationale in a comment.

Reply:
Btw, now that I think about it, it's pretty trivial to construct arbitrary namedtuple on the fly, using the factory function: https://docs.python.org/3/library/collections.html#collections.namedtuple

Reply:
ok, documented